File: src/java/tests/detailed/handler/GeolocationHandler.java
// Copyright (c) 2014 The Chromium Embedded Framework Authors. All rights
// reserved. Use of this source code is governed by a BSD-style license that
// can be found in the LICENSE file.

package tests.detailed.handler;

import java.awt.Frame;

import javax.swing.JOptionPane;
import javax.swing.SwingUtilities;

import org.cef.browser.CefBrowser;
import org.cef.callback.CefGeolocationCallback;
import org.cef.handler.CefGeolocationHandlerAdapter;

public class GeolocationHandler extends CefGeolocationHandlerAdapter {
  private final Frame owner_;
  public GeolocationHandler(Frame owner) {
    owner_ = owner;
  }

  @Override
  public boolean onRequestGeolocationPermission(CefBrowser browser,
                                             String requesting_url,
                                             int request_id,
                                             CefGeolocationCallback callback) {
    final CefGeolocationCallback cb = callback;
    final String reqUrl = requesting_url;
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        int dialogResult = JOptionPane.showConfirmDialog(owner_,
            "The URL \n" + reqUrl +
            "\nwants to request your geolocation." +
            "\nDo you want to proceed?",
            "Geolocation requested",
            JOptionPane.YES_NO_OPTION,
            JOptionPane.QUESTION_MESSAGE);
        cb.Continue(dialogResult == JOptionPane.YES_OPTION);
      }
    });
    return true;
  }
}

File: native/geolocation_handler.cpp
// Copyright (c) 2014 The Chromium Embedded Framework Authors. All rights
// reserved. Use of this source code is governed by a BSD-style license that
// can be found in the LICENSE file.

#include "geolocation_handler.h"
#include "client_handler.h"

#include "jni_util.h"
#include "util.h"

GeolocationHandler::GeolocationHandler(JNIEnv* env, jobject handler) {
  jhandler_ = env->NewGlobalRef(handler);
}

GeolocationHandler::~GeolocationHandler() {
  JNIEnv* env = GetJNIEnv();
  env->DeleteGlobalRef(jhandler_);
}

bool GeolocationHandler::OnRequestGeolocationPermission(
      CefRefPtr<CefBrowser> browser,
      const CefString& requesting_url,
      int request_id,
      CefRefPtr<CefGeolocationCallback> callback) {
  JNIEnv* env = GetJNIEnv();
  if (!env)
    return false;
  jobject jcallback =
      NewJNIObject(env, "org/cef/callback/CefGeolocationCallback_N");

  if (!jcallback)
    return false;
  SetCefForJNIObject(env, jcallback, callback.get(), "CefGeolocationCallback");
  jboolean jresult = JNI_FALSE;
  JNI_CALL_METHOD(env, jhandler_,
                  "onRequestGeolocationPermission",
                  "(Lorg/cef/browser/CefBrowser;Ljava/lang/String;"
                  "ILorg/cef/callback/CefGeolocationCallback;)Z",
                  Boolean,
                  jresult,
                  GetJNIBrowser(browser),
                  NewJNIString(env, requesting_url),
                  request_id,
                  jcallback);

  if (jresult == JNI_FALSE) {
    SetCefForJNIObject<CefGeolocationCallback>(env, jcallback, NULL,
        "CefGeolocationCallback");
  }
  return (jresult != JNI_FALSE);
}

void GeolocationHandler::OnCancelGeolocationPermission(
      CefRefPtr<CefBrowser> browser,
      int request_id) {
  JNIEnv* env = GetJNIEnv();
  if (!env)
    return;
  JNI_CALL_VOID_METHOD(env, jhandler_,
                       "onCancelGeolocationPermission",
                       "(Lorg/cef/browser/CefBrowser;I)V",
                       GetJNIBrowser(browser),
                       request_id);
}
